src/ReservoirPriceOracle.sol

```diff
@@ -350,43 +483,89 @@ contract ReservoirPriceOracle is IPriceOracle, IReservoirPriceOracle, Owned(msg.
         if (aToken1 < aToken0) revert RPC_TOKENS_UNSORTED();
-        if (aRoute.length > MAX_ROUTE_LENGTH || aRoute.length < 2) revert RPC_INVALID_ROUTE_LENGTH();
-        if (aRoute[0] != aToken0 || aRoute[aRoute.length - 1] != aToken1) revert RPC_INVALID_ROUTE();
+        if (lRouteLength > MAX_ROUTE_LENGTH || lRouteLength < 2) revert RPC_INVALID_ROUTE_LENGTH();
+        if (aRoute[0] != aToken0 || aRoute[lRouteLength - 1] != aToken1) revert RPC_INVALID_ROUTE();
 
-        _route[aToken0][aToken1] = aRoute;
-        emit Route(aToken0, aToken1, aRoute);
+        bytes32 lSlot = aToken0.calculateSlot(aToken1);
+
+        // simple route
+        if (lRouteLength == 2) {
+            uint256 lToken0Decimals = IERC20(aToken0).decimals();
+            uint256 lToken1Decimals = IERC20(aToken1).decimals();
+            if (lToken0Decimals > 18 || lToken1Decimals > 18) revert RPC_UNSUPPORTED_TOKEN_DECIMALS();
+
+            int256 lDiff = int256(lToken1Decimals) - int256(lToken0Decimals);
+
+            bytes32 lData = FlagsLib.FLAG_SIMPLE_PRICE.combine(lDiff);
+            assembly {
+                // Write data to storage.
+                sstore(lSlot, lData)
+            }
+        }
+        // composite route
+        else {
+            address lSecondToken = aRoute[1];
+            address lThirdToken = aRoute[2];
+            // Set the uppermost byte of lFirstWord to FLAG_COMPOSITE_NEXT for intermediate hops
+            bytes32 lFirstWord = FlagsLib.FLAG_COMPOSITE_NEXT << 248;
+            // Move the address to start on the 2nd byte.
+            bytes32 lSecondTokenData = bytes32(bytes20(lSecondToken)) >> 8;
+            bytes32 lThirdTokenFirst10Bytes = bytes32(bytes20(lThirdToken)) >> 176;
+
+            // Trim away the first 10 bytes since we only want the last 10 bytes.
+            bytes32 lThirdTokenSecond10Bytes = bytes32(bytes20(lThirdToken) << 80);
+            bytes32 lSecondWord;
+
+            if (lRouteLength == 3) {
+                // Set flag before third token to FLAG_COMPOSITE_END
+                lFirstWord = lFirstWord | lSecondTokenData | FlagsLib.FLAG_COMPOSITE_END << 80 | lThirdTokenFirst10Bytes;
+
+                lSecondWord = lThirdTokenSecond10Bytes;
+            } else if (lRouteLength == 4) {
+                // Set flag before third token to FLAG_COMPOSITE_NEXT as there are 4 tokens in total
+                lFirstWord =
+                    lFirstWord | lSecondTokenData | FlagsLib.FLAG_COMPOSITE_NEXT << 80 | lThirdTokenFirst10Bytes;
+
+                bytes32 lFourthTokenData = bytes32(bytes20(aToken1)) >> 88;
+                lSecondWord = lThirdTokenSecond10Bytes | FlagsLib.FLAG_COMPOSITE_END << 168 | lFourthTokenData;
+
+                _checkAndPopulateIntermediateRoute(lThirdToken, aToken1);
+            }
+            _checkAndPopulateIntermediateRoute(aToken0, lSecondToken);
+            _checkAndPopulateIntermediateRoute(lSecondToken, lThirdToken);
```

Comment:
I'm confused by this, assume the route:

A -> B -> C

Are AB and BC the intermediate routes? Why do we need to store them too?

Reply:
as explained in discord, resolving